server/app/rss.py

```diff
@@ -0,0 +1,101 @@
+import feedparser
+from .watchlist import Watchlist
+from .watchlist import unpack_entry
+from .watchlist import url_formats
+
+def get_rss_url(watchlist_item):
+	channel_id, source, site = unpack_entry(watchlist_item)
+	return url_formats[source]['template'].format(site, channel_id)
+
+def from_file(filename):
+	return feedparser.parse(filename)
+
+def from_watchlist_item(watchlist_item):
+	return feedparser.parse(get_rss_url(watchlist_item))
+
+def raw_feed(watchlist_path):
+	watchlist = Watchlist(watchlist_path)
+	feeds = list(map(from_watchlist_item, watchlist))
+	return feeds
+
+def _assign_path(dic, path, item):
+	cursor = dic
+	for key in path[:-1]:
+		if key in cursor:
+			cursor = cursor[key]
+		else:
+			cursor[key] = {}
+			cursor = cursor[key]
+	cursor[path[-1]] = item
+
+def _in_cursor(key, cursor):
+	if isinstance(cursor, list) or isinstance(cursor, tuple):
+		if type(key) is int:
+			return key >= 0 and key < len(cursor)
+	if isinstance(cursor, dict):
+		return key in cursor
+	return False #pragma: nocover
+
+def _flatten(target, source, target_path, source_path, optional=False):
+	if type(target_path) is int or type(target_path) is str:
+		target_path = (target_path,)
+	if type(source_path) is int or type(source_path) is str:
+		source_path = (source_path,)
+
+	cursor = source
+	for source_key in source_path:
+		correct_types = (list, tuple, dict)
+		is_correct_type = any(map(lambda t: isinstance(cursor, t), correct_types))
+		if not is_correct_type or not _in_cursor(source_key, cursor): #pragma: nocover
+			if not optional:
+				raise ValueError('invalid rss feed')
+			_assign_path(target, target_path, None)
+			return
+		cursor = cursor[source_key]
+	_assign_path(target, target_path, cursor)
+
+def summarize_feed(feed, raise_error=False):
+	summary = {}
+	def reword(summary_path, feed_path, optional=False):
+		_flatten(summary, feed, summary_path, feed_path, optional=optional)
+
+	try:
+		if feed.get('bozo', None) is not False:
+			raise ValueError('invalid rss feed')
+		reword(('channel', 'name'), ('feed', 'title'))
+		reword(('channel', 'id')  , ('feed', 'yt_channelid'))
+		reword(('channel', 'url') , ('feed', 'href'))
+
+		videos = ( summarize_entry(entry, raise_error=raise_error)
+					for entry in feed.get('entries', tuple()) )
+		videos = list(filter(lambda x: x is not None, videos))
+		summary['videos'] = videos
+
+	except ValueError as e:
+		if raise_error:
+			raise e
+		return None
+	return summary
+
+def summarize_entry(entry, raise_error=False):
+	summary = {}
+	def reword(summary_path, entry_path, optional=False):
```

Comment:
duplicate definition of the same function

Reply:
I understood by looking into it, but would also be great to get a little comment explaining what it does

Reply:
@Edmond120  Saw you made a commit in response to this. I was referring to `reword` function specifically, it's defined twice in `summarize_entry` and `summarize_feed`. We can make them into one function in the outer scope if we use more general argument names.

Reply:
About that, both reword definitions are closures and since they capture a different state, they are not really duplicates.
I will still make them into one function though since I've noticed that after a change _flatten() is no longer doing any flattening.

Reply:
Yeah they access (more like populate) the `summary` variable but the function can just as well define it inside and return it. That would actually be better since `reword` would become a pure function.

Reply:
@aziz512 done